File: python/openjiuwen/algorithm/passthrough.py
from __future__ import annotations

"""与 Rust passthrough.rs 同功能：选第一个可用目标。"""

from openjiuwen.contrib import PyAlgorithm


class Passthrough(PyAlgorithm):
    name = "passthrough"

    def decide(self, request, ctx):
        targets = list(getattr(ctx, "targets", []) or [])
        exclusions = set(getattr(request, "exclusions", []) or [])
        for model in targets:
            if model not in exclusions:
                return {
                    "selected_model_id": model,
                    "reasoning": "passthrough: first available target",
                    "is_answer_call": True,
                }
        raise ValueError("no available target after exclusions")

File: crates/algorithms/src/passthrough.rs
use openjiuwen_protocol::{Decision, RouteRequest, RouterError};

use crate::algorithm::{Algorithm, RouteContext};

/// 直通：选目标集中第一个未被排除的模型。
pub struct Passthrough;

impl Algorithm for Passthrough {
    fn name(&self) -> &str {
        "passthrough"
    }

    fn decide(&self, request: &RouteRequest, ctx: &RouteContext) -> Result<Decision, RouterError> {
        let available = ctx.targets.without(&request.exclusions);
        let model = available.first().ok_or(RouterError::NoTarget)?;
        Ok(Decision::answer(model, "passthrough: first available target"))
    }
}
